File: Engine/ceCore/entity/csTransform.cc

#include <ceCore/entity/csTransform.hh>
#include <ceCore/entity/csSpatialState.hh>

namespace cryo
{


csTransform::csTransform(csSpatialState *state, const Matrix4f &matrix, const Matrix4f &parentMatrix)
    : m_state(state), m_matrix(matrix), m_translation(0.0f, 0.0f, 0.0f), m_rotation(0.0f, 0.0f, 0.0f, 1.0f), m_scale(
    1.0f,
    1.0f,
    1.0f), m_dirty(true), m_parentMatrix(parentMatrix)
{
  ExtractTRS();
}

void csTransform::ExtractTRS()
{
  m_translation = m_matrix.GetTranslation();
  m_scale       = Vector3f(m_matrix.GetXAxis().Length(),
                           m_matrix.GetYAxis().Length(),
                           m_matrix.GetZAxis().Length());

  Matrix3f rotMat(m_matrix.m00 / m_scale.x, m_matrix.m01 / m_scale.x, m_matrix.m02 / m_scale.x,
                  m_matrix.m10 / m_scale.y, m_matrix.m11 / m_scale.y, m_matrix.m12 / m_scale.y,
                  m_matrix.m20 / m_scale.z, m_matrix.m21 / m_scale.z, m_matrix.m22 / m_scale.z
  );
  m_rotation = Quaternion::FromMatrix(rotMat);
}

csTransform &csTransform::SetTranslation(const Vector3f &translation)
{
  m_translation = translation;
  m_dirty       = true;
  return *this;
}

csTransform &csTransform::SetTranslation(float x, float y, float z)
{
  m_translation.x = x;
  m_translation.y = y;
  m_translation.z = z;
  m_dirty = true;
  return *this;
}

const Vector3f &csTransform::GetTranslation() const
{
  return m_translation;
}


csTransform &csTransform::SetRotation(const Quaternion &rotation)
{
  m_rotation = rotation;
  m_dirty    = true;
  return *this;
}

const Quaternion &csTransform::GetRotation() const
{
  return m_rotation;
}


csTransform &csTransform::SetScale(const Vector3f &scale)
{
  m_scale = scale;
  m_dirty = true;
  return *this;
}


const Vector3f &csTransform::GetScalar() const
{
  return m_scale;
}

csTransform &csTransform::LookAt(const Vector3f &at, const Vector3f &up)
{
  Vector3f from = Matrix4f::Transform(m_parentMatrix, m_translation);

  Matrix4f matrix;
  matrix.SetLookAtInv(from, at, up);

  Matrix4f parent = m_parentMatrix;
  parent.Invert();
  m_matrix = matrix * parent;

  ExtractTRS();
  m_dirty = true;
  return *this;
}


const Matrix4f &csTransform::GetMatrix() const
{
  UpdateMatrix();
  return m_matrix;
}


const Matrix4f &csTransform::GetGlobalMatrix() const
{
  UpdateMatrix();
  return m_globalMatrix;
}

Vector3f csTransform::GetForward() const
{
  UpdateMatrix();
  return m_globalMatrix * Vector3f(0, 0, 1);
}


Vector3f csTransform::GetBackward() const
{
  UpdateMatrix();
  return m_globalMatrix * Vector3f(0, 0, -1);
}


Vector3f csTransform::GetLeft() const
{
  UpdateMatrix();
  return m_globalMatrix * Vector3f(-1, 0, 0);
}


Vector3f csTransform::GetRight() const
{
  UpdateMatrix();
  return m_globalMatrix * Vector3f(1, 0, 0);
}


Vector3f csTransform::GetUp() const
{
  UpdateMatrix();
  return m_globalMatrix * Vector3f(0, 1, 0);
}


Vector3f csTransform::GetDown() const
{
  UpdateMatrix();
  return m_globalMatrix * Vector3f(0, -1, 0);
}

void csTransform::Finish()
{
  UpdateMatrix();
  m_state->SetLocalMatrix(m_matrix);
}


void csTransform::UpdateMatrix() const
{
  if (m_dirty)
  {
    m_matrix = m_rotation.ToMatrix4();
    m_matrix.SetTranslation(m_translation);
    m_matrix.Scale(m_scale);
    m_globalMatrix = m_parentMatrix * m_matrix;
    m_dirty        = false;
  }

}

}

File: JCryoStorm/CsCore/src/main/java/org/cryo/core/entity/CsTransform.java
package org.cryo.core.entity;

import org.cryo.core.math.CsMatrix3f;
import org.cryo.core.math.CsMatrix4f;
import org.cryo.core.math.CsQuaternion;
import org.cryo.core.math.CsVector3f;

public class CsTransform
{

    private final CsSpatialState spatialState;
    private final CsVector3f translation = new CsVector3f();
    private final CsQuaternion rotation = new CsQuaternion();
    private final CsVector3f scale = new CsVector3f(1, 1, 1);

    private final CsMatrix4f parentMatrix = new CsMatrix4f();
    private boolean dirty;
    private final CsMatrix4f matrix = new CsMatrix4f();
    private final CsMatrix4f globalMatrix = new CsMatrix4f();


    public CsTransform(CsSpatialState spatialState) {
        this.spatialState = spatialState;
    }


    public CsTransform set(CsMatrix4f matrix,
                           CsMatrix4f parentMatrix) {
        this.matrix.set(matrix);
        this.parentMatrix.set(parentMatrix);

        extractTRS();
        return this;
    }

    private void extractTRS() {
        matrix.getTranslation(this.translation);
        scale.set(matrix.getXAxis(new CsVector3f()).length(),
                  matrix.getYAxis(new CsVector3f()).length(),
                  matrix.getZAxis(new CsVector3f()).length());

        CsMatrix3f rotMat = new CsMatrix3f(matrix.m00 / scale.x, matrix.m01 / scale.x, matrix.m02 / scale.x,
                                       matrix.m10 / scale.y, matrix.m11 / scale.y, matrix.m12 / scale.y,
                                       matrix.m20 / scale.z, matrix.m21 / scale.z, matrix.m22 / scale.z);
        rotation.set(CsQuaternion.fromMatrix(rotMat));
    }

    public CsTransform setTranslation(CsVector3f translation) {
        this.translation.set(translation);
        dirty = true;
        return this;
    }

    public CsTransform setTranslation(float x, float y, float z) {
        translation.set(x, y, z);
        dirty = true;
        return this;
    }

    public CsVector3f getTranslation(CsVector3f out) {
        out.set(translation);
        return out;
    }


    public CsTransform setRotation(CsQuaternion rotation) {
        this.rotation.set(rotation);
        dirty = true;
        return this;
    }


    public CsQuaternion getTranslation(CsQuaternion out) {
        out.set(rotation);
        return out;
    }

    public CsTransform setScale(CsVector3f scale) {
        this.scale.set(scale);
        dirty = true;
        return this;
    }

    public CsTransform setScale(float x, float y, float z) {
        this.scale.set(x, y, z);
        dirty = true;
        return this;
    }

    public CsVector3f getScale(CsVector3f out) {
        out.set(scale);
        return out;
    }

    public CsTransform lookAt(CsVector3f at, CsVector3f up) {

        CsVector3f from = this.parentMatrix.transform(this.translation, new CsVector3f());


        CsMatrix4f tmpMatrix = new CsMatrix4f();
        tmpMatrix.setLookAtInv(from, at, up);

        CsMatrix4f parent = new CsMatrix4f(parentMatrix);
        parent.invert();

        tmpMatrix.mul(parent, this.matrix);

        extractTRS();
        dirty = true;
        return this;
    }


    public CsMatrix4f getMatrix() {
        updateMatrix();
        return matrix;
    }


    public CsMatrix4f getGlobalMatrix() {
        updateMatrix();
        return globalMatrix;
    }

    public CsVector3f getForward() {
        updateMatrix();
        return globalMatrix.mul(CsVector3f.Z, new CsVector3f());
    }

    public CsVector3f getBackward() {
        updateMatrix();
        return globalMatrix.mul(CsVector3f.NEG_Z, new CsVector3f());
    }

    public CsVector3f getLeft() {
        updateMatrix();
        return globalMatrix.mul(CsVector3f.NEG_X, new CsVector3f());
    }

    public CsVector3f getRight() {
        updateMatrix();
        return globalMatrix.mul(CsVector3f.X, new CsVector3f());
    }


    public CsVector3f getUp() {
        updateMatrix();
        return globalMatrix.mul(CsVector3f.Y, new CsVector3f());
    }

    public CsVector3f getDown() {
        updateMatrix();
        return globalMatrix.mul(CsVector3f.NEG_Y, new CsVector3f());
    }


    public void finish() {
        updateMatrix();
        this.spatialState.setLocalMatrix(this.matrix);
    }


    void updateMatrix() {
        if (dirty) {
            this.rotation.toMatrix4(this.matrix);
            this.matrix.setTranslation(this.translation);
            this.matrix.scale(this.scale);

            this.globalMatrix.set(parentMatrix).mul(this.matrix, globalMatrix);
            this.dirty = false;
        }
    }


}
